python/ray/serve/api.py

```diff
@@ -317,23 +318,38 @@ def shadow_traffic(self, endpoint_name: str, backend_tag: str,
                                                    proportion))
 
     @_ensure_connected
-    def get_handle(self, endpoint_name: str) -> RayServeHandle:
+    def get_handle(self,
+                   endpoint_name: str,
+                   missing_ok: Optional[bool] = False) -> RayServeHandle:
         """Retrieve RayServeHandle for service endpoint to invoke it from Python.
 
         Args:
             endpoint_name (str): A registered service endpoint.
+            missing_ok (bool): If true, then Serve won't check the endpoint is
+                registered. False by default.
 
         Returns:
             RayServeHandle
         """
-        if endpoint_name not in ray.get(
+        if not missing_ok and endpoint_name not in ray.get(
                 self._controller.get_all_endpoints.remote()):
             raise KeyError(f"Endpoint '{endpoint_name}' does not exist.")
 
-        # TODO(edoakes): we should choose the router on the same node.
-        routers = ray.get(self._controller.get_routers.remote())
+        routers = list(ray.get(self._controller.get_routers.remote()).values())
+        current_node_id = ray.get_runtime_context().node_id.hex()
+
+        try:
+            router_chosen = next(
+                filter(lambda r: get_node_id_for_actor(r) == current_node_id,
+                       routers))
+        except StopIteration:
+            logger.warning(
+                f"When getting a handle for {endpoint_name}, Serve can't find "
+                "a router on the same node. Serve will use a random router.")
+            router_chosen = random.choice(routers)
```

Comment:
this logic is a lot of "magic" -- any reason not to just iterate over the routers in a for loop? would be more readable

Reply:
I will just leave this logic as is. they will be gone really really soon with upcoming refactoring ;-)